### DISCUSSION TITLE
classes/class_resource

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#122] [User-contributed notes]
# classes/class_resource

Inherits: RefCounted< Object Inherited By: Animation, AnimationLibrary, AnimationNode, AnimationNodeStateMachinePlayback, AnimationNodeStateMachineTransition, AudioBusLayout, AudioEffect, AudioStre...

https://docs.godotengine.org/en/stable/classes/class_resource.html

**Comment by Hanekawa-chan** [comment]
Very important detail is that when you try loading resource that has _init(), it is going to break loading, so you should replace it with some custom constructor function

**Reply by Calinou** [reply]
It'll work if `_init()` doesn't take any parameters, but if it does, you can't insantiate it through the Create New Node/Resource dialog anymore. See https://github.com/godotengine/godot-proposals/issues/1513.

**Comment by steveweenie** [comment]
Is it a good idea to put gameplay features in the resource? Because it seems persistent. I've been implementing this resource based weapon system in multiplayer and splitscreen, but the equip functions and it's current ammo are in the resource. Whenever I queue free the gameplay scene and player character, then loading it back up it remembers how much ammo it previously had and that it was already equipped from the first match. I was thinking just putting these features in the node instead.

**Reply by RedMser** [reply]
Resources loaded from files are cached by default, so as long as they stay loaded, changes will be remembered. You can use "local to scene" or call .duplicate() on the resource to work around this.

**Comment by hungptsg** [comment]
Godot `.tres` are akin to `JSON` but for Godot. They can contain other `Resource`s, like a Texture2D, but no `Node`.

Great for making game-related static data.

`.tres` can be edit in the inspector.

**Comment by themipper** [comment]
Loading and Saving of Resources:

Something that is worth mentioning here. If you want to use ResourceLoader and ResourceSaver to save/load your Resource make sure that variables that you want to load and save have a @export annotation like @export or @export_storage. 
Otherwise they get ignored in the load and save functions.

**Comment by insomniacUNDERSCORElemon** [comment]
I'm trying to assign Godot materials in an import script, there really is no simple way to get the no-EXT no-path name (like I expect `.name`/ `.get_name()` would)? I don't want to need to edit `resource_name` for every (in-engine) material especially if I decide to change names.

I guess a tool script could be used to assign name to resource_name (and update if changed) though that feels needlessly complicated. Plus getting that name is not the most elegant either: `.resource_path.get_basename().get_slice("/", 4)`.

**Comment by Terpki** [comment]
2nd paragraph in the Description section: 
Shouldn't it say 
"In GDScript, resources can be* loaded from disk by their..."
 instead of 
"In GDScript, resources can loaded from disk by their..."?

**Comment by OS-of-S** [comment]
I think it should be noted, bocuse it took a time for me to understand it:

It seems that if 'local to scene' Resource is saved as exported property of other Resource, which is not local to scene, than nested Resource not local to scene even if an according tick is here. Sounds logical... Thus you, for exemple, cannot get root of local scene right - it will allways be a root of current scene.

<img width="615" height="597" alt="image" src="https://github.com/user-attachments/assets/564ef04f-eb7a-425e-b0e9-ef880c106b05" />

**Comment by Nodragem** [comment]
is it normal that NOTIFICATION_PREDELETE is not called on a resource when a node containing this resource is deleted?

**Comment by nenad-jalsovec** [comment]
There should be some flag that enables automatic emission of `changed` signal for all properties in custom resources. As things are now, if you have a resource with many properties you need to implement dummy setters on all properties just so each makes a call to `emit_changed()`.

**Reply by Calinou** [reply]
This is being tracked in this proposal:

- https://github.com/godotengine/godot-proposals/issues/10720
